### ISSUE TITLE
evaluation = 0

### ISSUE THREAD

**Issue by kcrumb** [google/automl#398]
I took [create_pascal_tfrecord.py](https://github.com/google/automl/blob/master/efficientdet/dataset/create_pascal_tfrecord.py) and modified it to [convert my custom data](https://github.com/kcrumb/automl/blob/master/efficientdet/dataset/create_polyp_tfrecord.py) to TFrecord. During training there are no errors but on evalution I always get a mAP of 0.
````
Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.000
Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.000
Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
````

Also my loss is jumping around.
![loss](https://user-images.githubusercontent.com/4051421/81732175-5e81fc80-9490-11ea-8eb5-a715b7b71cf7.png)

My command is:
````
!python /content/automl-efficientdet/efficientdet/main.py \
--mode="train_and_eval" \
--training_file_pattern="/path/p-*.tfrecord" \
--validation_file_pattern="/path/p-*.tfrecord" \
--val_json_file="/path/json_p.json" \
--model_name=efficientdet-d0 \
--model_dir="/content/logs/" \
--ckpt="/efficientdet-d0" \
--train_batch_size=4 \
--eval_batch_size=1 \
--eval_samples=1000 \
--num_examples_per_epoch=1000 \
--num_epochs=10 \
--hparams="num_classes=1,moving_average_decay=0,learning_rate=0.001,var_exclude_expr=r'.*/class-predict/.*'" \
--use_tpu=False
````
I tried it with and without `val_json_file`.
Am I missing something in my tfrecord convertion script? Or what what am I doing wrong?
Thanks in advance!

**Comment by mingxingtan** [comment]
It is hard to tell.  Maybe you can try to use a smaller learning rate?

**Comment by kcrumb** [comment]
Thx! I found my problem. We have to explicitly define the `background ID = 0` and set all other IDs > 0 (found the hint in #391). Furthermore set `num_classes=2` and I defined `learning_rate=0.002` and `lr_warmup_init=0.0002`. This worked for me and the loss is decreasing. I believe that the fluctuating loss is due to the low batch size, even `train_batch_size=16` appears to be too small. After setting `clip_gradients_norm=5.0` I was able to get even better values on the validation set as before. Maybe this will help someone in the future.

**Comment by shreymohan** [comment]
Hey @kcrumb , which optimizer did you use?

**Comment by kcrumb** [comment]
@shreymohan I'm using `sgd`. The `adam` optimizer didn't work well in my early tests, but probably we just have to play around with the parameters to get `adam` running properly.

**Comment by shreymohan** [comment]
Thank you for your reply! I have 1 more question. While fine-tuning on my custom dataset the eval MAP continuously decreases because I did not add background class while creating tfrecords and val.json. This is how my label_dict looked like : {'car':0,'van':1,'bus':2,'others':3}
I changed it to {'background':0,'car':1,'van':2,'bus':3,'others':4 } in my create_tfrecord script and would increase the num classes by 1. 

Is there anything else I need to change?

**Comment by kcrumb** [comment]
@shreymohan Not that I'm aware of. The the background label was my primary source of error. You just need to have a decent batch size because this could be another source of error, if the network doesn't learn properly. After that I just played around with the hyperparameters to see what fits best to my dataset.

**Comment by shreymohan** [comment]
Thank you @kcrumb, now I get decent evaluation map while using SGD  optimizer as the network learns. With Adam it was decreasing, don't know why. Thank you for your help, very grateful!
